File: labs/kernel/driver_entries/materials/sysfs_view.py
from pathlib import Path
import sys

def describe(entry):
    # sysfs 是活动视图：观察期间设备可能消失。
    try:
        location = entry.resolve(strict=True)
        driver = (entry / "driver").resolve(strict=True).name
    except FileNotFoundError:
        if not entry.exists():
            return f"{entry.name}: disappeared"
        location = entry.resolve()
        driver = "(no driver link)"
    return f"{entry.name}\n  object={location}\n  driver={driver}"

def main():
    root = Path(sys.argv[1]) if len(sys.argv) > 1 else Path("/sys/bus/platform/devices")
    if not root.is_dir():
        raise SystemExit(f"不是可读取的设备目录: {root}")
    entries = sorted(root.iterdir())[:5]
    if not entries:
        print("当前目录没有设备；这不表示内核中没有其他总线。")
    for entry in entries:
        print(describe(entry))

if __name__ == "__main__":
    main()

File: labs/kernel/driver_entries/materials/sysfs_view.cpp
// 只读视图：每次查询都有失败边界，不冻结设备拓扑。
#include <algorithm>
#include <filesystem>
#include <iostream>
#include <system_error>
#include <vector>

namespace fs = std::filesystem;

static void describe(const fs::path &entry)
{
    std::error_code error;
    const fs::path location = fs::canonical(entry, error);
    if (error) {
        std::cout << entry.filename().string() << ": object unavailable: "
                  << error.message() << '\n';
        return;
    }
    std::cout << entry.filename().string() << "\n  object="
              << location.string() << '\n';
    const fs::path driver = fs::canonical(entry / "driver", error);
    if (error == std::errc::no_such_file_or_directory)
        std::cout << "  driver=(no resolvable driver link)\n";
    else if (error)
        std::cout << "  driver unavailable: " << error.message() << '\n';
    else
        std::cout << "  driver=" << driver.filename().string() << '\n';
}

int main(int argc, char **argv)
{
    if (argc > 2) {
        std::cerr << "用法: sysfs_view [设备目录]\n";
        return 1;
    }
    const fs::path root = argc == 2 ? argv[1] : "/sys/bus/platform/devices";
    try {
        if (!fs::is_directory(root)) {
            std::cerr << "不是设备目录: " << root.string() << '\n';
            return 1;
        }
        std::vector<fs::path> entries;
        for (const auto &entry : fs::directory_iterator(root))
            entries.push_back(entry.path());
        std::sort(entries.begin(), entries.end());
        if (entries.empty())
            std::cout << "当前目录没有设备；这不表示内核中没有其他总线。\n";
        const std::size_t count = std::min(entries.size(), std::size_t{5});
        for (std::size_t index = 0; index < count; ++index)
            describe(entries[index]);
    } catch (const fs::filesystem_error &error) {
        std::cerr << "目录观察失败: " << error.what() << '\n';
        return 1;
    }
}
